src/v/kafka/server/connection_context.cc

```diff
@@ -571,6 +609,38 @@ connection_context::dispatch_method_once(request_header hdr, size_t size) {
     auto self = shared_from_this();
     auto rctx = request_context(
       self, std::move(hdr), std::move(buf), sres->backpressure_delay);
+
+    if (!_is_virtualized_connection || !rctx.header().client_id.has_value()) {
```

Comment:
i don't understand the purpose of the second part of this conditional. if it is virtualized and there isn't a client id should we shutdown the connection?

Reply:
yes this seems like a mistake?

Reply:
the idea was to fallback to standard processing whenever a client id is not provided

Reply:
> the idea was to fallback to standard processing whenever a client id is not provided

but isn't a connection either virtualized or not? i would recommend making the protocol strict and relax it later if needed. so i wonder if the following is true:

1. a connection is virtualized if and only if the first request received contains a magic client id
2. all requests received on a virtual connection have a client id that specifies their virtual connection

if there are requests that have no logical connection preference, then i guess that is the exception. can you document the protocol somewhere? that would be the easiest thing.

Reply:
Yeah I think if we expect the connection to be virtualized and then receive a request without a client ID, we should just drop the connection as it's misbehaving.

Reply:
>Yeah I think if we expect the connection to be virtualized and then receive a request without a client ID, we should just drop the connection as it's misbehaving.

Yep, subject to further guidance on the protocol being implemented, this would be my preference too.

Reply:
changed the logic here to drop connection if virtual connection id is invalid